### DISCUSSION TITLE
Can the input query complexity be dynamic based on actual input values?

### DISCUSSION THREAD

**Discussion by thosakwe** [Shopify/function-examples#566] [Feedback: General Improvements]
Hi Shopify Functions team,

I’d like to revisit the input query complexity limitation, particularly when merchants configure discounts with empty input values. For example, in a query like:

```graphql
query Input($collectionIds: [ID!]) {
  cart {
    lines {
      merchandise {
        __typename
        ...on ProductVariant {
          inCollections(ids: $collectionIds)
        }
      }
    }
  }
}

}
```

If the `collectionIds` array is empty, the input still gets penalized by 3 points, even though no additional data is being requested.

Would it be possible to base the query complexity on actual input values rather than the static query? It would make more sense if complexity was enforced when the function instance is created with its input values, rather than when the function is published. This would allow merchants to configure discounts without empty arrays or unneeded data points affecting query complexity.

This wouldn’t just help public apps—though they would benefit greatly by being able to handle more use cases, especially since most merchants using Shopify Functions rely on apps unless they're on Shopify Plus. Making the complexity dynamic based on actual inputs could enable a wider range of use cases without increasing server load, while still keeping the 30-point limit flexible and specific to what merchants need.

I understand this might be difficult to implement, but it could greatly improve flexibility across the Shopify Functions ecosystem.

Thank you for considering this!

**Comment by thosakwe** [comment]
I guess what I'm saying is that it would be great if each function instance could have its own input query instead of a static one for all. This change would let individual merchants customize functions based on different criteria, increasing flexibility for various use cases.

**Comment by HarmonyHarvest** [comment]
I've been needing the ability to enable Combination of Product Discount and Free Shipping in Same Code for about a year now. I've had to get scrappy with workarounds, but it would be extremely helpful for my business.

**Comment by thosakwe** [comment]
I have opened a thread about this in the Developer Community Forum: https://community.shopify.dev/t/workarounds-for-the-input-query-complexity-limitation/23896?u=tobebuilds
